### PR TITLE
Feat/water 3038

### PR DESCRIPTION
**Pull request by timtamimi**
+ Adds a bunch of new columns for storing the calculation variables (factors) from CM
+ Adds a small fix/tweak whereby invoice totals are calculated on the fly instead of fetched from CM (as those variables are no longer supplied by CM)
+ Populates factors values in transactions in the CM Refresh Service
+ Changes .map to Bluebird.mapSeries when fetching invoices, as per discussion with James and Alan about simultaneous requests

### CODE REVIEW

File: src/lib/models/validators.js [hunk 1]
```diff
@@ -78,6 +79,8 @@ const assertQuantityWithMaximum = (value, max) => assert(value, VALID_QUANTITY.m
 const assertNullableQuantity = value => assert(value, VALID_NULLABLE_QUANTITY);
 const assertIsoString = value => assert(value, VALID_ISO_DATE_STRING);
 const assertFactor = value => assert(value, VALID_FACTOR);
+const assertFactorWithPrefix = value => assert(value, VALID_FACTOR_WITH_PREFIX);
+const assertNumeric = value => assert(value, Joi.number().allow(0).allow(null));
```
Comment (Stephan-Hough) [inline src/lib/models/validators.js:83]: if a value can be null we add nullable i.e.  `const assertNullableNumeric = ...`

File: src/lib/models/validators.js [hunk 2]
```diff
@@ -119,6 +122,8 @@ exports.assertQuantityWithMaximum = assertQuantityWithMaximum;
 exports.assertNullableQuantity = assertNullableQuantity;
 exports.assertIsoString = assertIsoString;
 exports.assertFactor = assertFactor;
+exports.assertFactorWithPrefix = assertFactorWithPrefix;
+exports.assertNumeric = assertNumeric;
```
Comment (Stephan-Hough) [inline src/lib/models/validators.js:126]: change these to match the previous comment

**Review by Stephan-Hough — APPROVED** [review]

File: migrations/sqls/20210322162609-add-generating-to-batch-status-enum-up.sql [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+DROP TYPE water.batch_status;
+
+CREATE TYPE water.batch_status AS ENUM (
+	'processing',
+	'review',
+	'ready',
+	'error',
+	'sent',
+	'empty',
+	'generating'
```
Comment (james-siteclick) [inline migrations/sqls/20210322162609-add-generating-to-batch-status-enum-up.sql:16]: I'm not certain if the new `generating` status is needed here - possibly the enum could have remained as it was?

File: src/lib/models/transaction.js [hunk 4]
```diff
@@ -309,6 +309,78 @@ class Transaction extends Model {
...
+  set calcSucFactor (calcSucFactor) {
+    validators.assertNullableNumeric(calcSucFactor);
+    this._calcSucFactor = calcSucFactor;
+  }
+
+  get calcSucFactor () {
+    return this._calcSucFactor;
+  }
+
+  set calcS126Factor (calcS126Factor) {
```
Comment (james-siteclick) [inline src/lib/models/transaction.js:348]: For the S126 and S127 factors, could we map these to a number and use a basic numeric type?

File: src/lib/models/validators.js [hunk 5]
```diff
@@ -13,6 +13,7 @@ const VALID_DATE = Joi.string().regex(dateRegex).required();
 const VALID_NULLABLE_DATE = VALID_DATE.allow(null);
 const VALID_ACCOUNT_NUMBER = Joi.string().regex(/^[ABENSTWY][0-9]{8}A$/).required();
 const VALID_LICENCE_NUMBER = Joi.string().regex(/^[&()*-./0-9A-Z]+$/).required();
+const VALID_FACTOR_WITH_PREFIX = Joi.string().regex(/^.... x [0-9].[0-9]$/).allow(null);
```
Comment (james-siteclick) [inline src/lib/models/validators.js:16]: As above this wouldn't be needed if we mapped the CM response to a number

File: src/modules/billing/mappers/transaction.js [hunk 6]
```diff
@@ -240,7 +248,19 @@ const modelToChargeModule = (batch, invoice, invoiceLicence, transaction) => {
     batchNumber: batch.id,
     ...mapChargeElementToChargeModuleTransaction(transaction.chargeElement),
     ...mapLicenceToChargeElementTransaction(invoiceLicence.licence),
-    newLicence: transaction.isNewLicence
+    newLicence: transaction.isNewLicence,
+    calculation: {
```
Comment (james-siteclick) [inline src/modules/billing/mappers/transaction.js:252]: Is this supposed to be here?  This mapper is for mapping data to create the transaction initially in the CM

File: src/modules/billing/mappers/transaction.js [hunk 7]
```diff
@@ -267,7 +287,15 @@ const cmToModelMapper = createMapper()
   .map('deminimis').to('isDeMinimis')
   .map('newLicence').to('isNewLicence')
   .map('twoPartTariff').to('isTwoPartTariffSupplementary', mapIsTwoPartTariffSupplementary)
-  .map('transactionStatus').to('status', mapCMTransactionStatus);
+  .map('transactionStatus').to('status', mapCMTransactionStatus)
+  .map('calculation.WRLSChargingResponse.sourceFactor').to('calcSourceFactor')
+  .map('calculation.WRLSChargingResponse.seasonFactor').to('calcSeasonFactor')
+  .map('calculation.WRLSChargingResponse.lossFactor').to('calcLossFactor')
+  .map('calculation.WRLSChargingResponse.sucFactor').to('calcSucFactor')
+  .map('calculation.WRLSChargingResponse.abatementAdjustment').to('calcS126Factor', val => val ? `S126 x ${val}` : null)
```
Comment (james-siteclick) [inline src/modules/billing/mappers/transaction.js:295]: This is where I wonder if we could just map to a number?

File: src/modules/billing/services/cm-refresh-service.js [hunk 8]
```diff
@@ -96,7 +96,15 @@ const mapTransaction = (invoice, transactionMap, cmTransaction) => {
       .get(cmTransaction.id)
       .fromHash({
         isDeMinimis: cmTransaction.isDeminimis,
-        value: cmTransaction.chargeValue
+        value: cmTransaction.chargeValue,
+        calcSourceFactor: cmTransaction.calculation.WRLSChargingResponse.sourceFactor,
```
Comment (james-siteclick) [inline src/modules/billing/services/cm-refresh-service.js:100]: Maybe less repetition if `WRLSChargingResponse` was destructured from `cmTransaction` first

**Review by james-siteclick — CHANGES_REQUESTED** [review]

File: migrations/sqls/20210322162609-add-generating-to-batch-status-enum-up.sql (hunk 3 above)
Comment (timtamimi) [reply migrations/sqls/20210322162609-add-generating-to-batch-status-enum-up.sql]: Removed migration in `81e36d2b`

File: src/modules/billing/services/cm-refresh-service.js (hunk 8 above)
Comment (timtamimi) [reply src/modules/billing/services/cm-refresh-service.js]: Totally fair. I've done this.

File: src/modules/billing/mappers/transaction.js (hunk 6 above)
Comment (timtamimi) [reply src/modules/billing/mappers/transaction.js]: I've removed this. Thanks

**Review by james-siteclick — APPROVED** [review]

**Pull request merged**